[Refactoring] Add async wrapper refactoring action

hamishknight opened the pull request:
This allows an async alternative function to be created that forwards onto the user's completion handler function through the use of `withCheckedContinuation`/`withCheckedThrowingContinuation`.

rdar://77802486

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
+    OS << " " << tok::l_brace << " cont " << tok::kw_in << "\n";
+
+    // asyncFn(args...) {
+    addCallToAsyncMethod(FD, TopHandler);
```
At some point it would be nice to handle the completion handler not being the last parameter, in which case this would fail. Could we add an assert here so we catch it if that does ever happen?

I think a comment here would help a lot as well - ie. this is actually calling the *sync* function, with the trailing closure added in the rest of the function.

hamishknight replied:
Decided to generalise the printing logic here to fit both use cases

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
+    SmallVector<SmallString<8>, 2> SuccessParamRefs;
+    for (auto idx : indices(SuccessParams)) {
+      auto &Param = SuccessParams[idx];
+      SmallString<8> Name;
```
Could just add to paramnames/refs straight away rather than creating and then copying into the vector, but it doesn't really matter.

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
+      Name.append("res");
+
+      // If we have multiple success params, number them e.g res1, res2...
+      if (SuccessParams.size() > 1)
```
Honestly somewhat tempted to just use the implicit names here, but this is slightly better in that it allows the user to rename if they want...

hamishknight replied:
I think it's kind of nice to spell the error parameter differently, but I'm not too fussed either way

bnbarham replied:
May as well leave as is since you've already got it :)

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
+      SuccessParamNames.emplace_back(Name);
+
+      // If we need to unwrap the success param, insert a force unwrap.
+      if (TopHandler.shouldUnwrap(Param.getParameterType()))
+        Name.append("!");
```
I guess there's not really much more we can do here... Konrad had the idea of a `fatalError` if both a `nil`, which would at least allow for a better error message. I'll see if anyone else has a better idea.

ktoso replied:
Maybe I'm overdoing it with that idea -- probably the `!` is fine 🤔

hamishknight replied:
Given it's generated code, I think a `fatalError` is a little nicer as we can better describe what went wrong.

hamishknight replied:
Decided to go with a `guard` and `fatalError`, how does this look?

bnbarham replied:
Seems reasonable, if a bit overkill. At least this way it's obvious what the assumptions are. Could have one guard for all the optional params? Though I suppose it's not obvious *which* is nil in that case 🤷.

hamishknight commented on include/swift/IDE/RefactoringKinds.def:
```diff
@@ -60,6 +60,8 @@ CURSOR_REFACTORING(ConvertToAsync, "Convert Function to Async", convert.func-to-
 
 CURSOR_REFACTORING(AddAsyncAlternative, "Add Async Alternative", add.async-alternative)
 
+CURSOR_REFACTORING(AddAsyncWrapper, "Add Async Wrapper Function", add.async-wrapper)
```
I wanted to make it clear it was adding a function and not something like e.g an `async` block wrapper call, but yeah I think it's probably better to be consistent here.

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6592,6 +6747,46 @@ bool RefactoringActionAddAsyncAlternative::performChange() {
+    return true;
+
+  // Deprecate the synchronous function.
+  EditConsumer.accept(SM, FD->getAttributeInsertionLoc(false),
```
Ah, I was following the example in the radar :) I think it makes sense to drop this

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
+        OS << tok::r_brace << "\n";
+      break;
+    }
+    case HandlerType::RESULT: {
```
Ah, neat!

hamishknight committed:
[test] Fix a couple of incorrect FileCheck directives

hamishknight committed:
[AST] Add front() and back() methods to ParameterList

hamishknight committed:
[Refactoring] Add utility method for printing tuples

hamishknight committed:
[Refactoring] Add async wrapper refactoring action

This allows an async alternative function to be
created that forwards onto the user's completion
handler function through the use of
`withCheckedContinuation`/`withCheckedThrowingContinuation`.

rdar://77802486

hamishknight committed:
[Refactoring] Generalize addCallToAsyncMethod

Factor out an `addForwardingCallTo` method that
allows the printing of a call to the old completion
handler function, with a given replacement for the
completion handler arg. `addCallToAsyncMethod` then
calls this with an empty replacement for the handler,
removing it from the call.

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows
